language/move-vm/types/src/values/values_impl.rs

```diff
@@ -230,7 +231,85 @@ impl Container {
     }
 
     fn signer(x: AccountAddress) -> Self {
-        Container::Struct(Rc::new(RefCell::new(vec![ValueImpl::Address(x)])))
+        Container::Struct(
+            Rc::new(RefCell::new(vec![ValueImpl::Address(x)])),
+            Some(Type::Signer),
+        )
+    }
+
+    fn check_type(&self, ty: &Type) -> PartialVMResult<()> {
+        match (self, ty) {
+            (Self::VecAddress(_), _)
+            | (Self::VecBool(_), _)
+            | (Self::VecU128(_), _)
+            | (Self::VecU64(_), _)
+            | (Self::VecU8(_), _)
+            | (Self::Vec(_), Type::Vector(_)) => Ok(()),
```

Comment:
These don't seem correct to me. I think
- For the specialized vectors
  - We should check if `ty` is indeed a vector of this particular type
- For the general vector
  - We should check if 
    - `ty` is indeed a vector type with non-specialized element types
    - All elements have the right type <= I would be concerned if we actually do this though, since it makes passing a vector to a different function O(n)